public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx

```diff
@@ -1,71 +1,165 @@
+  const [dataIsStale, setDataIsStale] = useState(false);
+  const { flag: explain, setFlag: setExplain } = useFlag(lokiQueryEditorExplainKey);
+  const { flag: rawQuery, setFlag: setRawQuery } = useFlag(lokiQueryEditorRawQueryKey, true);
 
-  const onLegendChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
-    const nextQuery = { ...query, legendFormat: e.currentTarget.value };
-    onChange(nextQuery);
+  const query = getQueryWithDefaults(props.query);
+  // This should be filled in from the defaults by now.
+  const editorMode = query.editorMode!;
+
+  const onExplainChange = (event: SyntheticEvent<HTMLInputElement>) => {
+    setExplain(event.currentTarget.checked);
   };
 
-  const legendField = (
-    <div className="gf-form-inline">
-      <div className="gf-form">
-        <InlineFormLabel
-          width={6}
-          tooltip="Controls the name of the time series, using name or pattern. For example
-        {{hostname}} will be replaced with label value for the label hostname. The legend only applies to metric queries."
-        >
-          Legend
-        </InlineFormLabel>
-        <input
-          type="text"
-          className="gf-form-input"
-          placeholder="legend format"
-          value={query.legendFormat || ''}
-          onChange={onLegendChange}
-          onBlur={onRunQuery}
-        />
-      </div>
-    </div>
+  const onEditorModeChange = useCallback(
+    (newEditorMode: QueryEditorMode) => {
+      reportInteraction('grafana_loki_editor_mode_clicked', {
+        newEditor: newEditorMode,
+        previousEditor: query.editorMode ?? '',
+        newQuery: !query.expr,
+        app: app ?? '',
+      });
+
+      if (newEditorMode === QueryEditorMode.Builder) {
+        const result = buildVisualQueryFromString(query.expr || '');
+        // If there are errors, give user a chance to decide if they want to go to builder as that can lose some data.
+        if (result.errors.length) {
+          setParseModalOpen(true);
+          return;
+        }
+      }
+      changeEditorMode(query, newEditorMode, onChange);
+    },
+    [onChange, query, app]
   );
 
+  useEffect(() => {
+    setDataIsStale(false);
+  }, [data]);
+
+  const onChangeInternal = (query: LokiQuery) => {
+    setDataIsStale(true);
+    onChange(query);
+  };
+
+  const onQueryPreviewChange = (event: SyntheticEvent<HTMLInputElement>) => {
+    const isEnabled = event.currentTarget.checked;
+    setRawQuery(isEnabled);
+  };
+
   return (
-    <LokiQueryField
-      datasource={datasource}
-      query={query}
-      onChange={onChange}
-      onRunQuery={onRunQuery}
-      onBlur={onRunQuery}
-      history={[]}
-      data={data}
-      data-testid={testIds.editor}
-      range={range}
-      ExtraFieldElement={
-        <>
-          <LokiOptionFields
-            lineLimitValue={query?.maxLines?.toString() || ''}
-            resolution={query?.resolution || 1}
+    <>
+      <ConfirmModal
```

Comment:
If this is duplicated from `LokiQueryEditorSelector` I would seriously consider moving it to a component.

Reply:
Same old, same old. 🙂 The renaming of file made this more confusing than it is. I should have splitted it into 2 PRs.